packages/core/src/manifest.ts

```diff
@@ -38,27 +40,82 @@ function defaultManifest(): ManifestData {
   };
 }
 
-const manifestDir = '.openzeppelin';
+const MANIFEST_DEFAULT_DIR = '.openzeppelin';
+const MANIFEST_TEMP_DIR = 'openzeppelin';
+
+async function getDevelopmentInstanceId(provider: EthereumProvider, chainId: number): Promise<string | undefined> {
+  let hardhatMetadata;
+
+  try {
+    hardhatMetadata = await getHardhatMetadata(provider);
+  } catch (e: unknown) {
+    if (e instanceof Error && (e.message.includes('Method') || e.message.includes('method'))) {
+      // method not found
+      return undefined;
+    } else {
+      throw e;
+    }
+  }
+
+  if (hardhatMetadata.chainId !== chainId) {
+    throw new Error(
+      `Broken invariant: Hardhat metadata's chainId ${hardhatMetadata.chainId} does not match eth_chainId ${chainId}`,
+    );
+  }
+
+  return hardhatMetadata.instanceId;
+}
+
+function getSuffix(chainId: number, devInstanceId?: string) {
+  if (devInstanceId !== undefined) {
+    return `${chainId}-${devInstanceId}`;
+  } else {
+    return chainId;
+  }
+}
+
+async function makeAndGetTempDir() {
+  const tempDir = path.join(os.tmpdir(), MANIFEST_TEMP_DIR);
+  await fs.mkdir(tempDir, { recursive: true });
+  return tempDir;
+}
 
 export class Manifest {
   readonly chainId: number;
+  readonly devInstanceId: string | undefined;
   readonly file: string;
+  readonly manifestDir: string;
   private readonly fallbackFile: string;
 
   private locked = false;
 
   static async forNetwork(provider: EthereumProvider): Promise<Manifest> {
-    return new Manifest(await getChainId(provider));
+    const chainId = await getChainId(provider);
+    const devInstanceId = await getDevelopmentInstanceId(provider, chainId);
+    const dir = devInstanceId !== undefined ? await makeAndGetTempDir() : undefined;
+
+    return new Manifest(chainId, devInstanceId, dir);
   }
 
-  constructor(chainId: number) {
+  constructor(chainId: number, devInstanceId?: string, dir?: string) {
     this.chainId = chainId;
+    this.devInstanceId = devInstanceId;
+    this.manifestDir = dir ?? MANIFEST_DEFAULT_DIR;
+    debug('manifest dir:', this.manifestDir);
+
+    const fallbackName = `unknown-${getSuffix(chainId, devInstanceId)}`;
+    this.fallbackFile = path.join(this.manifestDir, `${fallbackName}.json`);
+
+    const networkName = networkNames[this.chainId];
+    if (networkName !== undefined && devInstanceId !== undefined) {
+      throw new Error(
+        `Broken invariant: development instance id ${devInstanceId} not expected for named network ${networkName}`,
```

Comment:
I don't think this is an invariant. Users can customize the chain id of their local Hardhat network and sometimes they do, in particular they make it match a known network. This may be used when forking that chain to test.

In those cases, we should use a temporary manifest file rather than writing to the real network manifest file.

Forking is important, and I would propose to tackle it separately, but this PR risks breaking current users' testing workflows.

Note that `hardhat_metadata` includes a `forkedNetwork` field. Maybe it's as simple as looking up the forked network manifest and copying it into the temporary manifest file as the starting point?

Reply:
Handled forking as well.
If using a forked network, reads from the real manifest unless a dev manifest already exists, then writes to dev manifest location.
If using a dev instance but non forked network, only reads from the dev manifest (even if a real manifest for the same chain id exists).